zypp/KeyManager.cc

```diff
@@ -0,0 +1,361 @@
+/*---------------------------------------------------------------------\
+|                          ____ _   __ __ ___                          |
+|                         |__  / \ / / . \ . \                         |
+|                           / / \ V /|  _/  _/                         |
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+#include "zypp/KeyManager.h"
+#include "zypp/PublicKey.h"
+#include "zypp/PathInfo.h"
+#include "zypp/base/Logger.h"
+#include "zypp/TmpPath.h"
+#include "zypp/base/String.h"
+
+#include <boost/thread/once.hpp>
+#include <boost/interprocess/smart_ptr/scoped_ptr.hpp>
+#include <gpgme.h>
+
+#include <stdio.h>
+
+// @TODO [threading]
+// make sure to call the init code of gpgme only once
+// this might need to be moved to a different location when
+// threads are introduced into libzypp
+boost::once_flag gpgme_init_once = BOOST_ONCE_INIT;
+using std::endl;
+
+//using boost::interprocess pointer because it allows a custom deleter
+typedef boost::interprocess::scoped_ptr<gpgme_data, boost::function<void (gpgme_data_t)>> GpgmeDataPtr;
+typedef boost::interprocess::scoped_ptr<_gpgme_key, boost::function<void (gpgme_key_t)>>  GpgmeKeyPtr;
+typedef boost::interprocess::scoped_ptr<FILE, boost::function<int (FILE *)>> FILEPtr;
+
+static void initGpgme ()
+{
+  gpgme_check_version(NULL);
+
+}
+
+//output operator from gpgme_error_t
+std::ostream & operator<<( std::ostream & str, const gpgme_error_t err )
+{
+  return str << "<"<<gpgme_strsource(err)<<"> "<<gpgme_strerror(err);
+}
```

Comment:
Did this work for you? `gpgme_error_t` is an integral type. My attempts to log a `gpgme_error_t err;` result in the plain number being written. A simple wrapper class could do the trick (similar to `zypp/base/Errno.h`):
```
namespace zypp
{
  struct GpgmeError
  {
    GpgmeError( gpgme_error_t err_r = GPG_ERR_NO_ERROR )
    : _err( err_r )
    {}
    operator gpgme_error_t() const { return _err; }
  private:
    const gpgme_error_t _err;
  };
  std::ostream & operator<<( std::ostream & str, const GpgmeError & obj )
  { return str << "<" << gpgme_strsource(obj) << "> " << gpgme_strerror(obj); }

  GpgmeError err;
  err = gpgme_op_...();
  if ( err )
      ERR << err << endl;
```

I was looking into this, because there are several locations, where errors occur, but are not logged. Not only gpgme errors, also e.g. a not existing signature file (`verifyAndReadSignaturesFprs`), etc.

In production environments we often can't ask the customer to try something different or to reproduce some error, so we should try to have all the info in the logfile we need to figure out what happened. E.g `verifyAndReadSignaturesFprs`: Whether `!PathInfo( signature ).isExist()` or `!dataFile` or... led to returning an empty list should be visible in the logfile. If we figure out we log too much we can still reduce it.

Reply:
Ok, I was not sure how verbose I should be in the code with error logging. I will go over the code again and put some more output in.

Reply:
Done